### PR TITLE
SNOW-1559348: Add Snowpark pandas TimedeltaIndex class

### PR DESCRIPTION
**Pull request by sfc-gh-nkumar**
Fixes SNOW-1559348

- Add TimedeltaIndex class to provide lazy implementation.
- Add documentation pages.
- Only supports construction and raises not implemented error for timedelta index specific methods and attributes.
- Also include type propagation fixes to query compiler methods `set_index_from_series`, `set_index_from_columns` and `drop`.

### CODE REVIEW

File: tests/integ/modin/index/test_timedelta_index_methods.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import pytest
+
+import snowflake.snowpark.modin.plugin  # noqa: F401
+from tests.integ.modin.sql_counter import sql_count_checker
+
+
+@sql_count_checker(query_count=3)
+def test_timedelta_index_construction():
+    # create from native pandas timedelta index.
+    index = native_pd.TimedeltaIndex(["1 days", "2 days", "3 days"])
```
Comment (sfc-gh-azhan) [inline tests/integ/modin/index/test_timedelta_index_methods.py:16]: just FYI, you can use pd.timedelta_range

**Review by sfc-gh-azhan — COMMENTED** [review]
Are we going to add doc in a different PR?

**Review by sfc-gh-azhan — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -62,6 +62,7 @@
 - Added support for `Series.dt.normalize` and `DatetimeIndex.normalize`.
 - Added support for `Index.is_boolean`, `Index.is_integer`, `Index.is_floating`, `Index.is_numeric`, and `Index.is_object`.
 - Added support for `DatetimeIndex.round`, `DatetimeIndex.floor` and `DatetimeIndex.ceil`.
+- Added support for lazy `TimedeltaIndex`.
```
Comment (sfc-gh-nkrishna) [inline CHANGELOG.md:65]: I think it makes more sense to add this a sub-point under the "Added limited support for `Timedelta` type including" a couple of lines above.

File: tests/integ/modin/index/test_timedelta_index_methods.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    assert isinstance(snow_index, pd.TimedeltaIndex)
+
+    # create from query compiler with timedelta type.
+    df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]}, index=index)
+    snow_index = df.index
+    assert isinstance(snow_index, pd.TimedeltaIndex)
+
+    # create from snowpark pandas timedelta index.
+    snow_index = pd.Index(pd.TimedeltaIndex([123]))
+    assert isinstance(snow_index, pd.TimedeltaIndex)
```
Comment (sfc-gh-nkrishna) [inline tests/integ/modin/index/test_timedelta_index_methods.py:27]: Now that `set_index_from_columns` should handle Timedelta, can we add a test for something like this:

```
date_range1 = pd.date_range("2000-01-01", periods=10, freq="h")
date_range2 = pd.date_range("2001-05-01", periods=10, freq="h")
date_range2 - date_range1
```

**Review by sfc-gh-nkrishna — COMMENTED** [review]
LGTM overall, left two comments

**Comment by sfc-gh-nkrishna** [comment]
@sfc-gh-nkumar Looks like there are some Build Doc failures that need addressing

File: tests/integ/modin/index/test_timedelta_index_methods.py (hunk 3 above)
Comment (sfc-gh-nkumar) [reply tests/integ/modin/index/test_timedelta_index_methods.py:27]: Ack

**Review by sfc-gh-nkrishna — APPROVED** [review]
Nice work! 🚢

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 4]
```diff
@@ -262,9 +262,14 @@ def __getattr__(self, key: str) -> Any:
...
-        return self.__constructor__(
-            self.to_series().reset_index(drop=True).__getattr__(method)(other)
-        )
+        series = self.to_series().reset_index(drop=True).__getattr__(method)(other)
+        qc = series._query_compiler
+        qc = qc.set_index_from_columns(qc.columns, include_index=False)
+        # Use base constructor to ensure that the correct type is returned.
+        idx = Index(query_compiler=qc)
+        if series.name is None:
+            idx.name = None
```
Comment (sfc-gh-nkrishna) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:271]: Why not do: `idx = Index(query_compiler=qc, name=series.name)`?

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:271]: I thinks that should also work, let me check.

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:271]: Oh, right now we don't support `name` argument in lazy index constructor.

Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py:271]: Maybe we can set it as `idx.name = series.name`?

Comment (sfc-gh-nkumar) [reply src/snowflake/snowpark/modin/plugin/extensions/index.py]: that worked!

**Pull request merged**